hudi-common/src/main/java/org/apache/hudi/common/table/read/buffer/ReusableKeyBasedRecordBuffer.java

```diff
@@ -0,0 +1,166 @@
+ * The implementation is used when point lookups are required for the same file group, such as the metadata table's files partition.
+ * As a result, the buffer also requires a set of valid keys to be provided to avoid returning all records in the processed logs when doing these point or small-batch lookups.
+ * @param <T> the engine specific record type
+ */
+public class ReusableKeyBasedRecordBuffer<T> extends FileGroupRecordBuffer<T> {
+  private final Set<String> validKeys;
+  private final Map<Serializable, BufferedRecord<T>> existingRecords;
+
+  ReusableKeyBasedRecordBuffer(HoodieReaderContext<T> readerContext, HoodieTableMetaClient hoodieTableMetaClient,
+                               RecordMergeMode recordMergeMode, PartialUpdateMode partialUpdateMode,
+                               TypedProperties props, Option<String> orderingFieldName,
+                               UpdateProcessor<T> updateProcessor, Map<Serializable, BufferedRecord<T>> records) {
+    super(readerContext, hoodieTableMetaClient, recordMergeMode, partialUpdateMode, props, orderingFieldName, updateProcessor);
+    this.existingRecords = records;
+    ValidationUtils.checkArgument(readerContext.getKeyFilterOpt().orElse(null) instanceof Predicates.In,
+        () -> "Key filter should be of type Predicates.In, but found: " + readerContext.getKeyFilterOpt().map(filter -> filter.getClass().getSimpleName()).orElse("NULL"));
+    List<Expression> children = ((Predicates.In) readerContext.getKeyFilterOpt().get()).getRightChildren();
+    this.validKeys = children.stream().map(e -> (String) e.eval(null)).collect(Collectors.toSet());
+  }
+
+  @Override
+  protected ExternalSpillableMap<Serializable, BufferedRecord<T>> initializeRecordsMap(String spillableMapBasePath) {
+    return null;
+  }
+
+  @Override
+  protected void initializeLogRecordIterator() {
+    logRecordIterator = new RemainingRecordIterator<>(validKeys, existingRecords);
+  }
+
+  @Override
+  public BufferType getBufferType() {
+    return BufferType.KEY_BASED_MERGE;
+  }
+
+  @Override
+  protected boolean doHasNext() throws IOException {
+    ValidationUtils.checkState(baseFileIterator != null, "Base file iterator has not been set yet");
+
+    // Handle merging.
+    while (baseFileIterator.hasNext()) {
+      if (hasNextBaseRecord(baseFileIterator.next())) {
+        return true;
+      }
+    }
+
+    // Handle records solely from log files.
+    return hasNextLogRecord();
+  }
+
+  protected boolean hasNextBaseRecord(T baseRecord) throws IOException {
+    String recordKey = readerContext.getRecordKey(baseRecord, readerSchema);
+    // Avoid removing from the map so the map can be reused later
+    BufferedRecord<T> logRecordInfo = existingRecords.get(recordKey);
+    if (logRecordInfo != null) {
+      validKeys.remove(recordKey);
+    }
+    return hasNextBaseRecord(baseRecord, logRecordInfo);
```

Comment:
For the reuse case with key filters, should the filter also apply on base file records, i.e., the semantics is that the record buffer should return the records correspond to keys in the predicate (`this.validKeys`), not to include all keys in the base file?  It seems that all the records have been read here.

Reply:
Or did I miss anything?

Reply:
You are missing the base file iterator portion of the FileGroupReader. That will generate an iterator over the base file with the filter provided by the reader context.